Studio-Lovelies/GG-JointJustice-Unity
# "Upload to itch.io"-step on `master`-branch pushes + minor packaging improvements

## Summary
### Major change: Added optional "upload to itch.io"-step to `master`-branch pushes

During my last PR the idea of [uploading release-versions of the game to itch.io instead of GitHub Releases came up](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/22#discussion_r693026156); so here it is! 👀

  **Note:**  
  Sadly this requires another one-time configuration of GitHub Secrets. Until the correct Secrets are added, the itch.io upload will simply be skipped.  
  Even with itch.io-Secrets in place, the **GitHub Releases will still be uploaded** in parallel.  
  To disable creation of GitHub Releases, just remove [line `101-149` from `/.github/workflows/build.yml`](https://github.com/ViMaSter/GG-JointJustice-Unity/blob/4ba50268b397b377f36aa4b27d1b1f16d352e334/.github/workflows/build.yml#L101-L149).

### Minor changes
### Packages  
- Added [the icon of the Lua-version as desktop icon](https://github.com/Studio-Lovelies/GameGrumpsJointJustice/blob/master/icon.png) to packaged builds
- Changed name of artifacts from `Standalone{platform}` to a more simple format (`Windows (x64)`, `macOS`, `Linux`, etc.)
- Downloaded `.zip`-files contain the GitHub Action Run number inside their filename (e.g. `Windows (x64) [#123].zip`)
- Packaged builds are now named `Game Grumps - Joint Justice`  
  ![image](https://user-images.githubusercontent.com/1689033/130336005-21d29f45-0c7d-46c0-a5eb-12882d0b7eeb.png)  
  (`:`, and thereby `Game Grumps: Joint Justice`, are not supported in file and folder names on most operating systems)
- When manually creating workflow runs, it is now possible to force an upload to GitHub Releases and/or itch.io, even when selecting non-`master`-branches  
 (entering anything but `false` [`true`, `1`, `yes`, empty] will trigger a full release)
  ![image](https://user-images.githubusercontent.com/1689033/130335343-3bf30335-aef2-40ca-a51c-231a806fcea0.png)
- Removed the `/UpgradeLog.htm`-file in the root of the repository from the latest .NET Upgrade

### Ingame  
- Packaged builds of the game no longer start in fullscreen, but windowed at `1280x768`
- Packaged builds of the game ([if built using a Unity Plus / Pro license](https://docs.unity3d.com/Manual/class-PlayerSettingsSplashScreen.html#:~:text=Subscription%20limitations,logo%20cannot%20be%20disabled.)) skip the Unity splash screen
  ![image](https://user-images.githubusercontent.com/1689033/130336305-15641d32-fed2-4c2c-9828-80de83cfa425.png)

## Before/after screenshots and/or animated gif
The itch.io-upload will require configuration to work. Otherwise it is simply skipped:
![image](https://user-images.githubusercontent.com/1689033/130335314-4299b48e-ded0-4702-bb5f-bdf63988d13a.png)

In order to upload to itch.io, three more [GitHub Secrets need to be added to this repository](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/settings/secrets/actions):
* `ITCHIO_USERNAME`
  When signing up to itch.io, you will be asked to select a username; **this will determine the URL for all projects** (i.e., as mine is `ViMaSter`, projects can be found at https://vimaster.itch.io/).  
  Enter this username (the value between `https://` and `.itch.io`) here.
* `ITCHIO_PROJECTNAME`
  [Create a new game on itch.io](https://itch.io/game/new). You will be asked to create a `Project URL`; enter the value **AFTER THE `/`** here (i.e. for `https://vimaster.itch.io/uploadtest` enter `uploadtest` as value for this secret)
* `ITCHIO_APIKEY`
  To generate this...
  1. create an account on itch.io
  2. sign in
  3. visit https://itch.io/user/settings/api-keys
  4. hit `Generate new API key`
  5. click `View` on the newly added row
  6. copy the full sequence of random characters
  7. add it as "value" of this secret

![image](https://user-images.githubusercontent.com/1689033/130334485-1fa54a25-a01a-4a31-83fa-15117356e9e1.png)

## Testing instructions
1. Push any change to `master`-branch
2. Wait for the build workflow to complete
3. Notice a new GitHub Release but **no itch.io build** being created
4. Follow the steps in `Before/after screenshots and/or animated gif`
5. Again push any change to the `master`-branch of this repository or manually queue a build and set the release trigger to `true` instead of `false`
6. Wait for the build workflow to complete
7. Notice a new GitHub Release being created
8. Visit your itch.io-project URL and notice **two Windows-, one macOS and one Linux build** being added to your project 
9. One more time push any change to the `master`-branch of this repository or manually queue a build and set the release trigger to `true` instead of `false`
10. Notice a new GitHub Release being created
11. Visit your itch.io-project URL again and notice an updated timestamp on the existing **two Windows-, one macOS and one Linux build**

**Regarding size and file limit concerns:** (Automated) Itch.IO uploads are [limited to 32GB](https://itch.io/t/46291/uploading-large-games-faq#:~:text=butler%20will%20let%20you%20upload%20a%2032GB%20game). I personally don't think there's a danger of running into that any time soon with a text adventure. :)

## Additional information
- `[ ]` Changes UI
- `[x]` Introduces new feature
- `[ ]` Removes existing feature
- `[x]` Has associated resource: 
  - https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/22#discussion_r693026156
